File: src/app_lib_wasm/buildMatrix.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <emscripten.h>



// Функция для обработки предвычисленных значений, полученных из JS
EMSCRIPTEN_KEEPALIVE
double *build_matrix_from_precomputed(double **precomputed_values, int rows, int cols, double *output_values, int output_length)
{
    if (!precomputed_values || rows <= 0 || cols <= 0 || !output_values || output_length <= 0)
    {
        printf("Invalid inputs to build_matrix_from_precomputed\n");
        return NULL;
    }

    // Выделяем память для матрицы
    double **matrix = (double **)malloc(rows * sizeof(double *));
    if (!matrix)
    {
        printf("Failed to allocate memory for matrix rows\n");
        return NULL;
    }

    // Выделяем память для каждой строки матрицы (rows + 1 для включения свободных членов)
    for (int i = 0; i < rows; i++)
    {
        matrix[i] = (double *)malloc((rows + 1) * sizeof(double));
        if (!matrix[i])
        {
            printf("Failed to allocate memory for matrix row %d\n", i);

            for (int j = 0; j < i; j++)
                free(matrix[j]);
            free(matrix);
            return NULL;
        }
        // Инициализируем нулями
        memset(matrix[i], 0, (rows + 1) * sizeof(double));
    }

    // Формируем матрицу системы линейных уравнений
    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < rows; j++)
        {
            double sum = 0.0;
            for (int k = 0; k < output_length; k++)
                sum += precomputed_values[i][k] * precomputed_values[j][k];

            matrix[i][j] = sum;
        }

        // Заполняем свободные члены (вектор b)
        double sum = 0.0;
        for (int k = 0; k < output_length; k++)
            sum += precomputed_values[i][k] * output_values[k];

        matrix[i][rows] = sum;
    }

    // Преобразуем двумерную матрицу в плоский массив
    double *flat_matrix = (double *)malloc(rows * (rows + 1) * sizeof(double));
    if (!flat_matrix)
    {
        for (int i = 0; i < rows; i++)
            free(matrix[i]);
        free(matrix);
        return NULL;
    }

    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < rows + 1; j++)
        {
            flat_matrix[i * (rows + 1) + j] = matrix[i][j];
        }
    }

    // Освобождаем двумерную матрицу
    for (int i = 0; i < rows; i++)
        free(matrix[i]);
    free(matrix);

    return flat_matrix;
}

File: src/app_lib_wasm/buildMatrix.js
import { getAppLib } from './moduleLoader';

export async function buildMatrixFromPrecomputed(precomputedValues, outputValues) {
  try {
    const wasm = await getAppLib();
    
    if (!precomputedValues || !precomputedValues.length || !outputValues || !outputValues.length) {
      console.error("Некорректные входные данные");
      return null;
    }
    
    const rows = precomputedValues.length;
    const cols = precomputedValues[0].length;

    const precomputedPtrs = [];
    for (let i = 0; i < rows; i++) {
      const rowPtr = wasm._malloc(cols * 8);
      new Float64Array(wasm.HEAPF64.buffer, rowPtr, cols).set(precomputedValues[i]);
      precomputedPtrs.push(rowPtr);
    }
    
    const precomputedArrayPtr = wasm._malloc(rows * 4);
    new Int32Array(wasm.HEAP32.buffer, precomputedArrayPtr, rows).set(precomputedPtrs);
    
    const outputPtr = wasm._malloc(outputValues.length * 8);
    new Float64Array(wasm.HEAPF64.buffer, outputPtr, outputValues.length).set(outputValues);
    
    
    const matrixPtr = wasm._build_matrix_from_precomputed(
      precomputedArrayPtr,
      rows,
      cols,
      outputPtr,
      outputValues.length
    );
    
    if (!matrixPtr) {
      console.error("Не удалось построить матрицу");
      
      for (let i = 0; i < rows; i++) wasm._free(precomputedPtrs[i]);

      wasm._free(precomputedArrayPtr);
      wasm._free(outputPtr);
      
      return null;
    }
    

    const flatMatrixPtr = wasm._matrix_to_flat(matrixPtr, rows, rows + 1);
    for (let i = 0; i < rows; i++) wasm._free(precomputedPtrs[i]);

    wasm._free(precomputedArrayPtr);
    wasm._free(outputPtr);
    wasm._free_matrix(matrixPtr, rows);
    
    if (!flatMatrixPtr) {
      console.error("Не удалось преобразовать матрицу в плоский массив");
      return null;
    }
    
    const resultMatrix = [];
    for (let i = 0; i < rows; i++) {
      const row = [];
      for (let j = 0; j <= rows; j++)
        row.push(wasm.HEAPF64[(flatMatrixPtr / 8) + i * (rows + 1) + j]);

      resultMatrix.push(row);
    }
    
    wasm._free(flatMatrixPtr);
    
    return resultMatrix;
    
  } catch (error) {
      console.error("Ошибка в buildMatrixFromPrecomputed:", error);
    return null;
  }
}

export async function computeMatrixWithC(precomputedValues, data) {
  
  if (!precomputedValues || !precomputedValues.length || !data || !data.length) {
    console.error('Некорректные входные данные');
    return null;
  }
  
  const outputKey = Object.keys(data[0])[0];
  const outputValues = data.map(item => item[outputKey]);
  
  return await buildMatrixFromPrecomputed(precomputedValues, outputValues);
}